constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,500 @@
+  // INVARIANT: If two adjacent intervals abut exactly (e.g. [0, 3), [3, 5]), their values are non-equal.
+  private final List<Segment<V>> segments;
+
+  public IntervalMap() {
+    this.segments = new ArrayList<>();
+    this.alg = new IntervalAlgebra(Interval.FOREVER);
+  }
+
+  public IntervalMap(final IntervalAlgebra algebra, final List<Segment<V>> segments) {
+    this(algebra);
+    for (final var segment: segments) {
+      this.setInternal(segment.interval(), segment.value(), 0);
+    }
+  }
+
+  public IntervalMap(final List<Segment<V>> segments) {
+    this.alg = new IntervalAlgebra();
+    this.segments = new ArrayList<>(segments.size());
+    for (final var segment: segments) {
+      this.setInternal(segment.interval(), segment.value(), 0);
+    }
+  }
+
+  public IntervalMap(final IntervalAlgebra algebra) {
+    this.alg = algebra;
+    this.segments = new ArrayList<>();
+  }
+
+  public IntervalMap(final Interval interval, final V value) {
+    this.alg = new IntervalAlgebra(Interval.FOREVER);
+    this.segments = List.of(Segment.of(interval, value));
+  }
+
+  @SafeVarargs
+  public IntervalMap(final Segment<V>... segments) {
+    this();
+    for (final var segment: segments) {
+      this.setInternal(segment.interval(), segment.value(), 0);
+    }
+  }
+
+  public IntervalMap(final IntervalAlgebra algebra, final Interval key, final V value) {
+    this(algebra, new ArrayList<>(List.of(Segment.of(Objects.requireNonNull(key), Objects.requireNonNull(value)))));
+  }
+
+  public IntervalMap(final List<Interval> intervals, final V value) {
+    this(
+        new IntervalAlgebra(Interval.FOREVER),
+        intervals.stream().map($ -> Segment.of($, value)).toList()
+    );
+  }
+
+  // Copy constructor
+  public IntervalMap(final IntervalMap<V> other) {
+    this(other.alg, new ArrayList<>(other.segments));
+  }
+
+  public IntervalMap(final V value) {
+   this(Interval.FOREVER, value);
+  }
+
+  public IntervalMap<V> set(final Interval interval, final V value) {
+    return set(new IntervalMap<>(interval, value));
+  }
+
+  public IntervalMap<V> set(final List<Interval> intervals, final V value) {
+    return set(new IntervalMap<>(intervals, value));
+  }
+
+  public IntervalMap<V> set(final IntervalMap<V> other) {
+    final var result = new IntervalMap<>(this);
+    int index = 0;
+    for (final var segment: other) {
+      index = result.setInternal(segment.interval(), segment.value(), index);
+    }
+    return result;
+  }
+
+  int setInternal(Interval interval, final V value, int index) {
+    interval = Interval.intersect(this.alg.bounds(), interval);
+    if (interval.isEmpty()) return index;
+
+    //truncate interval to the bottom
+    if (interval.start.shorterThan(this.alg.bottom().end)) {
+      interval = interval(
+          this.alg.bottom().end,
+          Inclusive,
+          interval.end,
+          interval.endInclusivity
+      );
+    }
```

Comment:
🔶 Why is this block necessary? It seems to me that the `nterval.intersect(this.alg.bounds(), interval)` above would subsume this truncation.

Reply:
Probably yes. I rewrote map2 on wednesday and removed `alg.bottom()`, so this block no longer exists. I hope to commit it soon , but I still have some tests to fix now that `alg.bottom()` is gone. Sorry :|

Reply:
I'm supposed to be working on M20 but I'll try to get my changes in today, to not delay the review process.